src/engine/database/mod.rs

```diff
@@ -134,6 +135,86 @@ pub trait DatabaseEndpoint {
     async fn pool(&self) -> Result<DatabasePool, Error>;
 }
 
+/// Represents the different type of comparison match operations
+#[derive(Clone, Debug)]
+pub enum Operation {
+    EQ,
+    CONTAINS,
+    IN,
+    GT,
+    GTE,
+    LT,
+    LTE
+}
+
+/// Struct representing a value comparison. In query operations, visitors take provided 
+/// operation/value nested map and converted them into a `Comparison` struct and pass
+/// it on the database-specific transaction for use in creating match portion of queries. 
+#[derive(Clone, Debug)]
+pub struct Comparison {
+    operation: Operation,
+    operand: Value,
+    negated: bool
+}
+
+impl Comparison {
+
+    pub fn new(operation: Operation, negated: bool, operand: Value) -> Self {
+        Comparison {
+            operation,
+            operand,
+            negated
+        }
+    }
+
+    pub fn default(v: Value) -> Self {
+        Self::new(Operation::EQ, false, v)
+    }
+
+}
+
+impl TryFrom<Value> for Comparison {
+
+    type Error = Error;
+
+    fn try_from(v: Value) -> Result<Comparison, Error> {
+        Ok(match v {
+            Value::String(_) => Comparison::default(v),
+            Value::Int64(_) => Comparison::default(v),
+            Value::Float64(_) => Comparison::default(v),
+            Value::Bool(_) => Comparison::default(v),
+            Value::Map(mut m) => {
+                let operation_str = m.keys().next()
+                    .ok_or(Error::ComparisonParsingFailed { message: "Comparison keys empty".to_string() })?
+                    .clone(); // TODO: is there a way to not clone here?
+                let operand = m.remove(&operation_str).unwrap();
```

Comment:
I'd recommend avoiding `unwrap` in production code. If the `into_iter` solution to obtain both key and value at the same time works, from the comment above, that solves this problem too, though.